### PR TITLE
[DRAFT] Toolbelt workflow

### PR DESCRIPTION
**Pull request by mlieberman85**
Do not merge.

There are some open questions here. This expects there to be a Goreleaser file along with. Dockerfile for the container release.

There are a few open issues to be resolved:
- Goreleaser requires a custom config for most things. How should we handle this in relation to the workflow?
- Which SBOM tools should we use? Ideally we should use a Go tool that can at the very least adequately list transitive dependencies including their relationships as opposed to just a flat SBOM of whatever shows up in the go.mod/go.sum
- How should we parameterize this?
- This includes a lot of stuff that might not be applicable to every Go project, e.g. generating a container. Should we make it conditional? Should the conditions be based on introspection of the project or based on parameters?
- There are other actions that should be included here like codeql and scorecard itself. However, those include scheduling.

To summarize there's a few generic items that need to be sorted:
- What tools (outside of OSSF) should we use?
- What if tools include additional config that can't fit into the workflow itself?
- How should we parameterize this?
- What assumptions about the intended output artifacts should we make?
- What do we do about dependent workflows that should run on a different schedule compared to the primary workflow

### CODE REVIEW

**Comment by mlieberman85** [comment]
On that last point: https://stackoverflow.com/questions/70093399/can-i-have-multiple-on-in-single-gitactions-workflow-file might be a potential solution.

**Comment by trishankatdatadog** [comment]
Tries to address https://github.com/ossf/toolbelt/issues/11

**Comment by mlieberman85** [comment]
https://github.com/mlieberman85/skootrs-updates-8/blob/main/.github/workflows/toolbelt.yml -- here's an example use

**Comment by MarkLodato** [comment]
As we discussed in our meeting today, my suggestion is to add documentation on what users would expect out of this and what the requirements are. That will help reviewers assess whether this implementation achieves the goals. It's fine if it's rough and not wordsmithed at this point - the intent is to convey to the reader what we expect this to solve and not solve.

**Comment by mlieberman85** [comment]
@MarkLodato I added a Readme and some docs. I hope this answers. If not I can update.

File: .github/workflows/README.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
+# Toolbelt Workflow
+
+Goal: A mechanism that implements as many OpenSSF best practices using OpenSSF tools as possible in a way that is simple to use for project maintainers and easy for end users to check projects using this mechanism are following best practices and creating important security metadata (e.g. SBOMs, in-toto SLSA, etc.)
```
Comment (marcelamelara) [inline .github/workflows/README.md:3]: As I understood it, the goal goes beyond "simple use" and is really about requiring minimal changes to existing CI/CD-based builds. If so, the text should probably be a bit more explicit about this. Also, on the end user side, how will the workflow actually help users check projects adherence to best practices? Is seeing that they use the workflow enough of a signal, or is there more they'd have to do to gain assurances?

File: .github/workflows/README.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+Secondary Goals:
+
+- Highlight where there are gaps in OpenSSF and related practices, frameworks, and tools, or their adoption, especially in different ecosystems (e.g. Sigstore signatures are not supported in Maven)
+- Highlight when implementing these practices or adopting the tools is difficult
+Show examples following these practices and implementing the tools 
+
+Outcomes:
+
+- Maintainers should be able to use this mechanism to implement OpenSSF practices and adopt OpenSSF tooling, e.g. add SLSA to their build and turn on Scorecard.
+- Maintainers should be able to easily know where data is generated, e.g. where SBOMs, SLSA, SAST reports, etc. live so they can use it for later analysis and understanding supply chain risks.
```
Comment (marcelamelara) [inline .github/workflows/README.md:14]: This is might be too complex for the first version of the workflow, but I think it would be great if the workflow incorporated some standard policy checking at certain stages. Maybe it's enough to even start with something basic like signature checking, but the idea is to further reduce extra analysis that maintainers have to do.

File: .github/workflows/README.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+- End users should also be able to easily know where data is generated, e.g. where SBOMs, SLSA, SAST reports, etc. live so they can use it to make decisions on risk as well as analysis later on when vulnerabilities or supply chain incidents happen.
+
+Requirements:
+
+- Should be easy to update (e.g. dependency bot should not complain about a dozen things for you to update every other day)
+- Should consist of pieces that easily interoperate.
+- Should be as simple as possible for the user as possible
+
+Proposed MVP Solution: A Github workflow-based solution for Go with Go being widely supported for most of the stuff we need
```
Comment (marcelamelara) [inline .github/workflows/README.md:23]: Nit
```suggestion
Proposed MVP Solution: A Github Actions workflow-based solution for Go projects with Go being widely supported by most of the tools we're including
```

**Comment by trishankatdatadog** [comment]
This is great, thanks! Now that developers are signing, how should they and their customers be expected to verify signatures? Should we also provide instructions for that? 🙂

File: .github/workflows/README.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+Secondary Goals:
+
+- Highlight where there are gaps in OpenSSF and related practices, frameworks, and tools, or their adoption, especially in different ecosystems (e.g. Sigstore signatures are not supported in Maven)
+- Highlight when implementing these practices or adopting the tools is difficult
+Show examples following these practices and implementing the tools 
+
+Outcomes:
+
+- Maintainers should be able to use this mechanism to implement OpenSSF practices and adopt OpenSSF tooling, e.g. add SLSA to their build and turn on Scorecard.
```
Comment (lumjjb) [inline .github/workflows/README.md:13]: nit: inconsistency with using parenthesis for wrapping (e.g. ...)

File: .github/workflows/README.md (hunk 1 above)
Comment (lumjjb) [reply .github/workflows/README.md:3]: +1 on defining this a bit better. Also better defining if this is meant to be used as a reference architecture, playbook or aimed for production use have fairly different implications on the design decisions in the toolbelt project.

File: .github/workflows/README.md (hunk 4 above)
Comment (lumjjb) [inline .github/workflows/README.md:13]: Maintainers and End Users are used in the context of the document, which  seem to me like they are referring to the same persona? (pardon my ignorance)

File: .github/workflows/README.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+- Maintainers should be able to easily know where data is generated, e.g. where SBOMs, SLSA, SAST reports, etc. live so they can use it for later analysis and understanding supply chain risks.
+- End users should also be able to easily know where data is generated, e.g. where SBOMs, SLSA, SAST reports, etc. live so they can use it to make decisions on risk as well as analysis later on when vulnerabilities or supply chain incidents happen.
+
+Requirements:
+
+- Should be easy to update (e.g. dependency bot should not complain about a dozen things for you to update every other day)
+- Should consist of pieces that easily interoperate.
+- Should be as simple as possible for the user as possible
```
Comment (lumjjb) [inline .github/workflows/README.md:21]:
```suggestion
- Should be as simple as possible for the user
```
nit

File: .github/workflows/README.md [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+Outcomes:
+
+- Maintainers should be able to use this mechanism to implement OpenSSF practices and adopt OpenSSF tooling, e.g. add SLSA to their build and turn on Scorecard.
+- Maintainers should be able to easily know where data is generated, e.g. where SBOMs, SLSA, SAST reports, etc. live so they can use it for later analysis and understanding supply chain risks.
+- End users should also be able to easily know where data is generated, e.g. where SBOMs, SLSA, SAST reports, etc. live so they can use it to make decisions on risk as well as analysis later on when vulnerabilities or supply chain incidents happen.
+
+Requirements:
+
+- Should be easy to update (e.g. dependency bot should not complain about a dozen things for you to update every other day)
```
Comment (lumjjb) [inline .github/workflows/README.md:19]: perhaps this could be generalized to low end user maintenance requirement? I suspect that if they upgrade their versions of goreleaser or other tooling they'd also have to interact with changing their configurations, which would fall under the same level of toil.

File: .github/workflows/README.md (hunk 1 above)
Comment (mlieberman85) [reply .github/workflows/README.md:3]: @lumjjb I think in this case a lot of it "depends" -- Ideally we would love to have this be something that is used in production by anyone who has a use case supported by the toolbelt. The issue is finding folks who would use this.

We've gotten folks who have said they're interested but actually getting them to use it is a different story.

File: .github/workflows/README.md (hunk 2 above)
Comment (mlieberman85) [reply .github/workflows/README.md:14]: Yes, I think that would be something on the roadmap. I was even thinking this could set up a basic in-toto layout.

File: .github/workflows/README.md (hunk 4 above)
Comment (mlieberman85) [reply .github/workflows/README.md:13]: Let me update the definition. It is maybe better to replace end users with consumers. So Maintainers of the project and consumers of that project will have some overlapping use cases and some non-overlapping use cases.

File: .github/workflows/README.md (hunk 1 above)
Comment (mlieberman85) [reply .github/workflows/README.md:3]: Notes from meeting:
- We want this to be in CI/CD since that is the most critical part of the SDLC for stopping malicious stuff from getting in.
- This workflow should ideally focus in Github actions but could expand beyond that.
- Ideally we want this used as the de facto build workflow.
- Here is an example: https://github.com/mlieberman85/skootrs-updates-8/blob/main/.github/workflows/toolbelt.yml

File: .github/workflows/README.md (hunk 4 above)
Comment (mlieberman85) [reply .github/workflows/README.md:13]: I'll fix this

File: .github/workflows/toolbelt.yaml [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+      - name: Login to GitHub Container Registry
+        uses: docker/login-action@e92390c5fb421da1463c202d546fed0ec5c39f20 # v3.1.0
+        with:
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+      - name: Set up Go
+        uses: actions/setup-go@0c52d547c9bc32b1aa3301fd7a9cb496313a4491 # v5.0.0
+        with:
+          go-version: "1.22"
```
Comment (marcelamelara) [inline .github/workflows/toolbelt.yaml:37]: The Go version may or may not be something else to parameterize.

Comment (mlieberman85) [reply .github/workflows/toolbelt.yaml:37]: I wonder if this is something we can pull in based on the go.mod or similar.
